internal/beater/config/agentconfig.go

```diff
@@ -23,36 +23,89 @@ import (
 	"fmt"
 	"strings"
 	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/apm-server/internal/elasticsearch"
+	"github.com/elastic/elastic-agent-libs/config"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/go-ucfg"
 )
 
-// KibanaAgentConfig holds remote agent config information
-type KibanaAgentConfig struct {
-	Cache Cache `config:"cache"`
+const msgInvalidConfigAgentCfg = "invalid value for `apm-server.agent.config.cache.expiration`, only accepting full seconds"
+
+// AgentConfig configuration for dynamically querying agent configuration
+// via Elasticsearch or Kibana.
+type AgentConfig struct {
+	ESConfig *elasticsearch.Config
+	Cache    Cache `config:"cache"`
+
+	es *config.C
+}
+
+func (c *AgentConfig) Unpack(in *config.C) error {
+	type agentConfig AgentConfig
+	cfg := agentConfig(defaultAgentConfig())
+	if err := in.Unpack(&cfg); err != nil {
+		return errors.Wrap(err, "error unpacking config")
+	}
+	*c = AgentConfig(cfg)
+
+	es, err := in.Child("elasticsearch", -1)
+	if err == nil {
+		c.es = es
+	} else {
+		var ucfgError ucfg.Error
+		if !errors.As(err, &ucfgError) || ucfgError.Reason() != ucfg.ErrMissing {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *AgentConfig) setup(log *logp.Logger, outputESCfg *config.C) error {
+	if float64(int(c.Cache.Expiration.Seconds())) != c.Cache.Expiration.Seconds() {
+		return errors.New(msgInvalidConfigAgentCfg)
+	}
+	if outputESCfg != nil {
+		log.Info("using Elasticsearch output config for fetching agent config")
+		if err := outputESCfg.Unpack(&c.ESConfig); err != nil {
+			return errors.Wrap(err, "unpacking Elasticsearch output config into agent config")
+		}
+	}
+	if c.es != nil {
+		if err := c.es.Unpack(&c.ESConfig); err != nil {
+			return errors.Wrap(err, "unpacking user supplied Elasticsearch config into agent config")
+		}
+	}
```

Comment:
The idea is that APM UI will only have to inject an API Key, and the server will use that along with the same URL/TLS/etc. config for querying ES.

Agreed that merging a different URL over the top would not be sensible, but I also don't think there's a use case for changing the URL? We could make it so only auth-related config is allowed under `apm-server.agent.config.elasticsearch`.

I also don't think we even need to document `apm-server.agent.config.elasticsearch` in apm-server.yml, since it's really only relevant for Fleet. For standalone APM Server, users just need to ensure the `output.elasticsearch.api_key` or username/password has sufficient privileges.

Reply:
I'd be in favour of removing everything else than the credentials (api-key, username, password) and adding a comment explaining that this is exposed to allow fine tuned permission control and required when working with Fleet.

Reply:
Only allowing api-key, username and password in `apm-server.agent.config.elasticsearch` and removed `apm-server.agent.config.elasticsearch` in `apm-server.yml` and `apm-server.docker.yml`. Using an exported field name so that no code is needed for unpacking. The downside is that not only the merged ESConfig is exposed but the input `apm-server.agent.config.elasticsearch` is exposed to the rest of the code as well. Please let me know if we would like custom unpack code and unexported field instead.

Reply:
I spent some time thinking about it. There is value in not exposing the input config and only exposing the merged config. This will avoid confusion for other code users. I've added back the custom unpack code and added tests to make sure the behavior remains the same.